Feature/auto complete

fahall opened the pull request:
* Adds UnregisterConsoleCommands for bulk deregistration
* Adds [AutoComplete] support 
* Downgrades Roslyn to 4.11. Using 4.13 currently gives me problems with Godot (running 4.4). The downgrade solves the issue for now.

ryan-linehan commented:
Does Roslyn only affect you when you are working on the plugin? Or when you add it as a nuget package to your project as well?

ryan-linehan commented on Limbo.Console.Generator/AutoCompletes.cs:
```diff
@@ -0,0 +1,70 @@
+
+        private static void ValidateAutoCompletes(string methodName, IEnumerable<AutoCompleteDefinition> autoCompletes)
+        {
+            var duplicateArgIndexes = autoCompletes
```
Should we also validate that a [ConsoleCommand] is present on the same member?

ryan-linehan replied:
If limbo console doesn't throw errors from the code generated with it on something without [ConsoleCommand] I'm not sure I care either way. If it does throw errors then I think we should check this. What do you think?

ryan-linehan commented on Limbo.Console.Generator/AutoCompletes.cs:
```diff
@@ -0,0 +1,70 @@
+                                      .Select(g => g.Key)
+                                      .ToList();
+
+            if (!duplicateArgIndexes.Any())
```
```suggestion
            if (!duplicateArgIndexes.Any())
               return
```

fahall commented:
> Does Roslyn only affect you when you are working on the plugin? Or when you add it as a nuget package to your project as well?

This seems to not be causing me problems any more 🤷 . Removed the downgrade. 

> default should handle this right?
Yes... my editorconfig style is set to prefer explicit. Removed for compliance with this project's preferred style.

> returns without extra conditions 
Updated to match preferred style.

> Should we also validate that a [ConsoleCommand] is present on the same member as the [AutoComplete] attribute?

Yes, we should. Working on that.

fahall commented:
🤦 Past Alex had already resolved the issue by only parsing autocompletes on `[ConsoleCommand]` methods in the first place.

I pushed my validation code anyways, since it separates concerns a bit better, but the `MustDecorateAConsoleCommand` validation is essentially unreachable.

If you don't like the added complexity, I'll just revert those changes.

fahall commented:
I've hit the Roslyn version issue again.

I don't fully understand the issue, and it's likely something specific to my project. For now I'm leaving this as 4.14. I'll open a separate issue if I cannot resolve my CICD build issues

ryan-linehan commented:
Looks good! Going to merge it in and will post a version update to the repo shortly to push it out to nuget. Thanks again for contributing!
